module/zfs/dmu_send.c

```diff
@@ -672,20 +692,47 @@ do_dump(dmu_sendarg_t *dsa, struct send_block_record *data)
 		arc_buf_t *abuf;
 		int blksz = dblkszsec << SPA_MINBLOCKSHIFT;
 		uint64_t offset;
+		enum zio_flag zioflags = ZIO_FLAG_CANFAIL;
+
+		/*
+		 * If we have large blocks stored on disk but the send flags
+		 * don't allow us to send large blocks, we split the data from
+		 * the arc buf into chunks.
+		 */
+		boolean_t split_large_blocks =
+		    data->datablkszsec > SPA_OLD_MAXBLOCKSIZE &&
+		    !(dsa->dsa_featureflags & DMU_BACKUP_FEATURE_LARGE_BLOCKS);
+		/*
+		 * We should only request compressed data from the ARC if all
+		 * the following are true:
+		 *  - stream compression was requested
+		 *  - we aren't splitting large blocks into smaller chunks
+		 *  - the data won't need to be byteswapped before sending
+		 *  - this isn't an embedded block
+		 *  - this isn't metadata (if receiving on a different endian
+		 *    system it can be byteswapped more easily)
+		 */
+		boolean_t request_compressed =
+		    (dsa->dsa_featureflags & DMU_BACKUP_FEATURE_COMPRESSED) &&
+		    !split_large_blocks && !BP_SHOULD_BYTESWAP(bp) &&
+		    !BP_IS_EMBEDDED(bp) && !DMU_OT_IS_METADATA(BP_GET_TYPE(bp));
```

Comment:
I brought this up with @dankimmel before. `split_large_blocks` is determined by `data->datablkszsec > SPA_OLD_MAXBLOCKSIZE` above (if the large block feature is not set) and `data->datablkszsec` is only determined by the  dnode's `dn_datablkszsec`. What this means is that we determine whether to do a compressed send based on whether the block _could_ be too large, instead of whether it actually is. The result is that lots of compressible blocks are sent uncompressed. I was able to replicate this behavior with this script. Adding -L to `zfs send` will allow it to properly send compressed data.

```
zpool create -f pool sdb
zfs create -o dedup=verify -o compression=on pool/test

for i in `seq 1 10`
do
    yes | dd of=/pool/test/file-$i oflag=sync bs=8192 count=1
done

zfs snapshot pool/test@now
zfs send -cp pool/test@now | zstreamdump -v
```